### Prepare & Process Dataset
- Identify data type
- Call back what process need in the next step regarding to the 10 question test case
- Prepare datset for: <br>
    1. Analysis
    2. Create Model and Evaluation
    3. Predict next month

In [1]:
# import library needed
import pandas as pd
import numpy as np
import re

In [2]:
# load dataset
df_raw = pd.read_csv("telco_customer_churn.csv")

# identify data type
df_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


In [3]:
# display top 5 dataframe
df_raw.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges Churn  
0           Electronic check          29.85         29.85    No  
1               Mailed check          56.95        1889.5    No  
2               Mailed check          53.85        108.15   Yes  
3  Bank transfer (automatic)          42.30       1840.75    No  
4           Electronic check          70.70        151.65   Yes  

[5 rows x 21 columns]

In [4]:
# transform column suppose to be numeric
print('''
TotalCharges suppose to be numeric value, float
''')

df_raw['TotalCharges'] = pd.to_numeric(df_raw['TotalCharges'], errors='coerce')
df_raw.info()


TotalCharges suppose to be numeric value, float

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   objec

In [5]:
# fill NaN value in TotalCharges with mean
df_raw['TotalCharges'].fillna(df_raw['TotalCharges'].mean(), inplace=True)
print("Sum if ane null value for each columns: ")
df_raw.isnull().sum()

Sum if ane null value for each columns: 


customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

In [6]:
# save df_raw
df_raw.to_csv("df_raw_prepared.csv", index=False)

### Data type of dataset was ready for further analysis

### Call Back:

#### ***9: Buatlah machine learning model untuk memprediksi pelanggan yang akan churn dibulan berikutnya dan tunjukan hasil evaluasi model tersebut***

#### ***10: Jelaskan Fitur apa saja yang paling berpengaruh pada model anda dan berapa banyak pelanggan yang akan churn di bulan berikutnya***


#### Breakdown and Plan

> Breakdown:
- Untuk membuat model machine learning dan melakukan evaluasi performa, dibutuhkan dataset untuk training dan testing.
- Untuk melakukan prediksi bulan berikutnya, apabila tidak disediakan data test, kita perlu membuat hold out dataset atau unseen dataset.
    
> Plan:
- Data train dan testing kita sebut sebagai data ML development, porsinya sekitar 80-90% dari raw data
- Sedangkan unseen dataset (data bulan berikutnya) sekitar 10-20% dari raw data, dengan komposisi label yang seimbang (misalnya 15 churn, 15 no churn) sehingga kita akan dapat mengukur bagaimana performa model memprediksi data baru diluar machine learning development.

### Unseen Dataset

In [7]:
# Identify number of labels

df_procc = df_raw.copy()
df_procc = df_procc.sort_values(by=['Churn'])
print("Count Labels")
df_procc['Churn'].value_counts()

Count Labels


No     5174
Yes    1869
Name: Churn, dtype: int64

In [8]:
df_procc.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure  \
0     7590-VHVEG  Female              0     Yes         No       1   
4334  3969-GYXEL  Female              0      No         No      11   
4333  5325-UWTWJ    Male              0     Yes         No      31   
4332  4194-FJARJ  Female              0     Yes        Yes      54   
4331  5743-KHMNA    Male              0      No         No      71   

     PhoneService     MultipleLines InternetService       OnlineSecurity  ...  \
0              No  No phone service             DSL                   No  ...   
4334          Yes                No     Fiber optic                  Yes  ...   
4333          Yes               Yes     Fiber optic                  Yes  ...   
4332          Yes               Yes             DSL                  Yes  ...   
4331          Yes               Yes              No  No internet service  ...   

         DeviceProtection          TechSupport          StreamingTV  \
0                      No                   No                   No   
4334                   No                   No                   No   
4333                   No                  Yes                   No   
4332                   No                  Yes                   No   
4331  No internet service  No internet service  No internet service   

          StreamingMovies        Contract PaperlessBilling  \
0                      No  Month-to-month              Yes   
4334                   No  Month-to-month               No   
4333                   No  Month-to-month              Yes   
4332                   No        Two year               No   
4331  No internet service        Two year               No   

                  PaymentMethod MonthlyCharges  TotalCharges  Churn  
0              Electronic check          29.85         29.85     No  
4334           Electronic check          76.40        838.70     No  
4333    Credit card (automatic)          90.55       2929.75     No  
4332  Bank transfer (automatic)          60.00       3273.95     No  
4331  Bank transfer (automatic)          25.55       1898.10     No  

[5 rows x 21 columns]

In [9]:
df_procc.tail()

customerID  gender  SeniorCitizen Partner Dependents  tenure  \
5441  3512-IZIKN  Female              0     Yes         No      70   
1161  5868-CZJDR    Male              0      No        Yes       1   
1159  4086-YQSNZ  Female              1     Yes         No       3   
1163  0135-NMXAP  Female              0      No         No      12   
2974  1481-ZUWZA    Male              0      No         No      28   

     PhoneService     MultipleLines InternetService OnlineSecurity  ...  \
5441          Yes               Yes             DSL            Yes  ...   
1161           No  No phone service             DSL             No  ...   
1159          Yes                No     Fiber optic             No  ...   
1163          Yes               Yes     Fiber optic             No  ...   
2974          Yes               Yes     Fiber optic            Yes  ...   

     DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
5441              Yes         Yes          No              No        Two year   
1161               No          No          No              No  Month-to-month   
1159               No          No         Yes              No  Month-to-month   
1163               No          No         Yes              No  Month-to-month   
2974               No          No          No             Yes  Month-to-month   

     PaperlessBilling              PaymentMethod MonthlyCharges  TotalCharges  \
5441               No    Credit card (automatic)          65.30       4759.75   
1161              Yes               Mailed check          31.35         31.35   
1159              Yes           Electronic check          80.40        224.05   
1163              Yes  Bank transfer (automatic)          89.75       1052.40   
2974              Yes    Credit card (automatic)          94.50       2659.40   

      Churn  
5441    Yes  
1161    Yes  
1159    Yes  
1163    Yes  
2974    Yes  

[5 rows x 21 columns]

In [10]:
# take 15 rows randomly for each labels
churn_yes = df_procc[df_procc['Churn']=="Yes"]
churn_no = df_procc[df_procc['Churn']=="No"]

unseen_yes = churn_yes.sample(n=15)
unseen_no = churn_no.sample(n=15)

unseen_dataset = pd.concat([unseen_yes,unseen_no], ignore_index=True)
print(unseen_dataset.shape)
unseen_dataset

(30, 21)


customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0   4847-QNOKA  Female              0      No         No       1          Yes   
1   5103-MHMHY  Female              0      No        Yes       1          Yes   
2   1455-ESIQH    Male              0      No         No       1          Yes   
3   8784-CGILN  Female              0      No         No      18          Yes   
4   9821-POOTN    Male              0     Yes         No      35          Yes   
5   2321-OMBXY  Female              0     Yes        Yes      38          Yes   
6   2984-MIIZL    Male              0      No         No       4          Yes   
7   3254-YRILK    Male              1      No         No      19          Yes   
8   6646-QVXLR    Male              1     Yes         No      62          Yes   
9   1704-NRWYE  Female              1      No         No       9          Yes   
10  0137-OCGAB  Female              0      No         No       1          Yes   
11  4844-JJWUY  Female              1      No         No       1          Yes   
12  0207-MDKNV  Female              0      No         No      15          Yes   
13  2433-KMEAS    Male              0      No         No      22          Yes   
14  2626-URJFX    Male              0     Yes        Yes       4          Yes   
15  5061-PBXFW  Female              0     Yes        Yes      32          Yes   
16  5076-YVXCM    Male              0      No         No      51          Yes   
17  2761-XECQW    Male              1     Yes         No       8           No   
18  3824-RHKVR  Female              0      No         No      17          Yes   
19  8221-HVAYI    Male              0     Yes        Yes      35          Yes   
20  8316-BBQAY  Female              0      No         No      65           No   
21  9086-YJYXS    Male              0     Yes        Yes      34          Yes   
22  4801-KFYKL    Male              0      No        Yes       8          Yes   
23  7542-CYDDM    Male              0      No         No      18          Yes   
24  4578-PHJYZ    Male              0     Yes        Yes      52          Yes   
25  6480-YAGIY    Male              0      No         No      45          Yes   
26  8605-ITULD  Female              0      No         No       1          Yes   
27  3982-DQLUS    Male              1     Yes        Yes      65          Yes   
28  5914-XRFQB    Male              0     Yes         No      72          Yes   
29  6892-BOGQE  Female              0     Yes         No      71          Yes   

       MultipleLines InternetService       OnlineSecurity  ...  \
0                 No             DSL                   No  ...   
1                 No             DSL                   No  ...   
2                 No     Fiber optic                   No  ...   
3                Yes     Fiber optic                   No  ...   
4                Yes     Fiber optic                   No  ...   
5                 No             DSL                  Yes  ...   
6                 No     Fiber optic                   No  ...   
7                Yes     Fiber optic                   No  ...   
8                Yes     Fiber optic                   No  ...   
9                 No     Fiber optic                   No  ...   
10               Yes     Fiber optic                   No  ...   
11                No     Fiber optic                   No  ...   
12                No     Fiber optic                  Yes  ...   
13               Yes             DSL                  Yes  ...   
14               Yes             DSL                   No  ...   
15                No             DSL                  Yes  ...   
16               Yes     Fiber optic                   No  ...   
17  No phone service             DSL                   No  ...   
18                No              No  No internet service  ...   
19               Yes             DSL                  Yes  ...   
20  No phone service             DSL                  Yes  ...   
21               Yes             DSL    

In [11]:
# save unseen dataset as a next month dataset
unseen_dataset.to_csv("unseen_dataset.csv", index=False)

### Data train and test (Machine learning development)

In [12]:
# define list of id unseen dataset (with condition ID must be unique in the raw dataset)
list_id_unseen = unseen_dataset['customerID'].tolist()

# select raw dataset if customerID not in list_id_unseen
df_development = df_raw[~df_raw['customerID'].isin(list_id_unseen)]
print(df_development.shape)
print("Number of row dataset compare with raw dataset about: ", round(len(df_development)/len(df_raw), 5)*100, "%")

df_development.head()

(7013, 21)
Number of row dataset compare with raw dataset about:  99.574 %


customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges  Churn  
0           Electronic check          29.85         29.85     No  
1               Mailed check          56.95       1889.50     No  
2               Mailed check          53.85        108.15    Yes  
3  Bank transfer (automatic)          42.30       1840.75     No  
4           Electronic check          70.70        151.65    Yes  

[5 rows x 21 columns]

In [13]:
# save data train and data test for further anlisys and development
df_development.to_csv("df_development.csv", index=False)

### Result & Reveal

1. Number of rows of raw dataset about 7043, we saved the dataframe in csv named "df_raw_prepared.csv"
2. Number of rows of unseeen dataset about 30, we take a few rows just because we need more data training and testing **(regarding to the imbalance class of churn label, will not beneficial if under represented label too much reduced)**. Slicing 30 rows with 50:50 proportion class is enough to simulate as a next month dataset. We saved the dataframe in csv named "unseen_dataset.csv"
3. Number of rows of data train+data test about 7013, we will split this dataframe into 75-80% as training set and the rest as testing/model evaluation set. We saved the dataframe in csv named "df_development.csv"